components/gitpod-protocol/src/util/scrubbing.ts

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { createHash } from "crypto";
+
+const redactedFields = ["auth_", "password", "token", "key", "jwt", "secret", "email"];
+const hashedFields = ["contextURL", "workspaceID", "username"];
```

Comment:
We have userID and workspaceID in LogContext. Would make sense to add `contextURL` there as well and then redact those.

Reply:
no need for logging the userName

Reply:
An idea was to have generic scrubbing which can we reuse between different clients. One of them is logging. Logging API is for sure can be improved to reduce the surface, although I see just console.log is used a lot.

Reply:
We can have an additional generic scrubbing layer if that eases some concerns. The primary way to make sure we don't log sensitive data should be enforced through guidelines and code reviews primarily. Also, typed logging context API helps guide developers logging the right things. Are we establishing guidelines for logging (and other signals)? That would help us to also take special care during code reviews. Things like `console` use could be checked automatically.
I would love to not rely on generic scrubbing too much, because I believe it gives a false sense of security.

Reply:
FWIW I think it would be good to search through the code and fix some of the log statements

Reply:
> The primary way to make sure we don't log sensitive data should be enforced through guidelines and code reviews primarily. Also, typed logging context API helps guide developers logging the right things. Are we establishing guidelines for logging (and other signals)? That would help us to also take special care during code reviews.

We decided to go with the approach of a generic layer first as we saw that as the easiest way to at least try to scrub all logs. I agree that having a typed logging context API would be really lovely but we found it to bee too heavy a lift to get in place for the dedicated beta. You can find the original scrubber PR [here](https://github.com/gitpod-io/gitpod/pull/17664) and more context around the longer term plans for logs [here](https://www.notion.so/gitpod/Data-Sanitisation-4747f7941ed54b28b2459bab5220febd?pvs=4#5cee6025ac15479bbea4525b7a0d0d00). For a high-level overview of our approach to data scrubbing see [here](https://www.notion.so/gitpod/Data-Sanitisation-4747f7941ed54b28b2459bab5220febd?pvs=4#1c0b9f2f49f7419ca6514c396b8961b9) or this [loom](https://www.notion.so/gitpod/Data-Sanitisation-4747f7941ed54b28b2459bab5220febd?pvs=4#8acc0e416089430ea8a559e820972ef7)

Reply:
@svenefftinge @mads-hartmann It would be good if we agree on something before I continue here. I can also refactor all clients if there is someone to review them. Again I'm not sure what we do with 3rd party deps using console.log.

Reply:
Yeah, I can check with Chris and Milan, but I'm very much in favor of going with structured logging primarily. That said we can proceed with this PR as it could still be good as a secondary sanitization. It should maybe warn loudly if something sensitive is logged outside of the structured data.